### ISSUE TITLE
Memory overflow when on publisher faster than Netty subscriber

### ISSUE THREAD

**Issue by aliubimov** [ReactiveX/RxNetty#333]
Hi, 

I've write simple server that publishing numbers to network. 

```
    final RxServer<ByteBuf, ByteBuf> tcpServer = RxNetty.createTcpServer(4555,
            newConnection ->

                    Observable.range(0, 10_000_000).flatMap(integer -> {
                        final ByteBuf out = newConnection.getAllocator().buffer(16);

                        out.writeBytes((integer.toString() + "\n").getBytes());
                        newConnection.writeBytes(out);

                        return newConnection.flush();
                    }).finallyDo(newConnection::close)
    );
```

As client I am using simple Unix nc: `nc 192.168.1.2 4555 > out.

Client reach about 1-2m messages and stops execution. At this time server starts aggressive garbage collection.

```
[GC pause (G1 Evacuation Pause) (young) 613M->4967K(1024M), 0.0037673 secs]
[GC pause (G1 Evacuation Pause) (young) 612M->6501K(1024M), 0.0036726 secs]
[GC pause (G1 Evacuation Pause) (young) 612M->6916K(1024M), 0.0035126 secs]
[GC pause (G1 Evacuation Pause) (young) 612M->215M(1024M), 0.0952821 secs]
[GC pause (G1 Evacuation Pause) (young) 327M->310M(1024M), 0.1077812 secs]
[GC pause (G1 Evacuation Pause) (young) 350M->349M(1024M), 0.0168655 secs]
[GC pause (G1 Evacuation Pause) (young) 391M->389M(1024M), 0.0146687 secs]
[GC pause (G1 Evacuation Pause) (young) 431M->429M(1024M), 0.0148156 secs]
[GC pause (G1 Evacuation Pause) (young) 471M->470M(1024M), 0.0142044 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 546M->531M(1024M), 0.0222533 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0082144 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.3462373 secs]
[GC remark, 0.0026009 secs]
[GC cleanup 589M->589M(1024M), 0.0011590 secs]
[GC pause (G1 Evacuation Pause) (young) 689M->636M(1024M), 0.0403816 secs]
[GC pause (G1 Evacuation Pause) (young) 800M->735M(1024M), 0.0417374 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 845M->800M(1024M), 0.0297947 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0127174 secs]
[GC concurrent-mark-start]
[GC concurrent-mark-end, 0.5333336 secs]
[GC remark, 0.0034470 secs]
[GC cleanup 862M->862M(1024M), 0.0011776 secs]
[GC pause (G1 Evacuation Pause) (young) 872M->849M(1024M), 0.0206853 secs]
[GC pause (G1 Evacuation Pause) (young) 893M->879M(1024M), 0.0133139 secs]
[GC pause (G1 Evacuation Pause) (young) (initial-mark) 921M->904M(1024M), 0.0119625 secs]
[GC concurrent-root-region-scan-start]
[GC concurrent-root-region-scan-end, 0.0029878 secs]
[GC concurrent-mark-start]
[GC pause (G1 Evacuation Pause) (young) 946M->933M(1024M), 0.0113852 secs]
[GC concurrent-mark-end, 0.6050126 secs]
[GC remark, 0.0036208 secs]
[GC cleanup 956M->956M(1024M), 0.0013235 secs]
[GC pause (G1 Evacuation Pause) (young) 975M->963M(1024M), 0.0109574 secs]
```

I think flatMap cause overload of IndexedRingBuffer, however I am not able to figure out why this happens and why netty stops sending data at all.

Is there any way to setup back-pressure ?

**Comment by aliubimov** [comment]
Looks like issue caused by improper usage of RxNetty by my application.

**Comment by NiteshKant** [comment]
@aliubimov Thanks for reporting the issue. Yeah it does look like `flatmap` buffer bloat, possibly due to lot of `flush()` pending. Here is a mission control screenshot showing the same:

![image](https://cloud.githubusercontent.com/assets/820005/6243803/5e10eb28-b6fe-11e4-8a0e-634139ae263c.png)

`SpscArrayQueue` is used by `RxRingBuffer` used in `flatmap`. However, as you mentioned, it is because of the flush slowing down. I will dig more into it.

> > > Is there any way to setup back-pressure ?

Unfortunately no but I am working on the same for milestone [0.5.0](https://github.com/ReactiveX/RxNetty/milestones/0.5.0).
For this simplistic case, you can buffer writes using standard rxjava buffer operator with one flush every buffer, that would perform better.

**Comment by aliubimov** [comment]
Hi, 

You are right, it is probably due to lot of `flish()` for small messages. There is ab netty bug report for this https://github.com/netty/netty/issues/1759

**Comment by NiteshKant** [comment]
Backpressure is implemented in 0.5.0, closing this issue.

**Comment by damoshushu** [comment]
@NiteshKant 
Could you please provide tool name you used?
[https://github.com/ReactiveX/RxNetty/issues/333#issuecomment-74825873](url)
